In [1]:
import pandas as pd
import numpy as np
import scipy
import os 

import matplotlib.pyplot as plt


from sklearn.naive_bayes import GaussianNB 
from sklearn.linear_model import LogisticRegression 
from sklearn import ensemble 
from sklearn.ensemble import RandomForestClassifier 
from sklearn.metrics import brier_score_loss, precision_score, recall_score,f1_score, roc_auc_score, accuracy_score 
from sklearn.metrics import confusion_matrix, roc_curve

from sklearn.preprocessing import StandardScaler 
from sklearn.feature_extraction import DictVectorizer
from sklearn.cluster import KMeans

import random

from scipy.stats import ttest_ind

In [2]:
part_D_prescriber_dir = os.path.join(".", "datasets", "Medicare_Part_D_Prescribers_by_Provider_2021.csv")

**Loading the part D dataset about precription drug provider from the CSV file.**

In [3]:
partD_pd = pd.read_csv(part_D_prescriber_dir)

In [5]:
partD_pd.head(5)

PRSCRBR_NPI Prscrbr_Last_Org_Name Prscrbr_First_Name Prscrbr_MI  \
0   1003000126             Enkeshafi            Ardalan        NaN   
1   1003000142                Khalil             Rashid        NaN   
2   1003000167               Escobar              Julio          E   
3   1003000175         Reyes-Vasquez            Belinda        NaN   
4   1003000423               Velotta           Jennifer          A   

  Prscrbr_Crdntls Prscrbr_Gndr Prscrbr_Ent_Cd                 Prscrbr_St1  \
0            M.D.            M              I   6410 Rockledge Dr Ste 304   
1            M.D.            M              I  4126 N Holland Sylvania Rd   
2             DDS            M              I              5 Pine Cone Rd   
3          D.D.S.            F              I     322 N Azusa Ave Ste 202   
4            M.D.            F              I            11100 Euclid Ave   

  Prscrbr_St2 Prscrbr_City  ... Bene_Male_Cnt  Bene_Race_Wht_Cnt  \
0         NaN     Bethesda  ...         153.0              231.0   
1   Suite 220       Toledo  ...         105.0              144.0   
2         NaN       Dayton  ...          14.0               32.0   
3         NaN    La Puente  ...           NaN                NaN   
4         NaN    Cleveland  ...           NaN               57.0   

   Bene_Race_Black_Cnt  Bene_Race_Api_Cnt Bene_Race_Hspnc_Cnt  \
0                 32.0               47.0                20.0   
1                143.0                NaN                 NaN   
2                  0.0                0.0                 NaN   
3                  0.0                NaN                 NaN   
4                  NaN                0.0                 NaN   

  Bene_Race_Natind_Cnt Bene_Race_Othr_Cnt Bene_Dual_Cnt  Bene_Ndual_Cnt  \
0                  0.0               28.0         157.0           201.0   
1                  0.0                NaN         166.0           133.0   
2                  0.0                NaN           0.0            35.0   
3                  0.0                0.0           NaN             NaN   
4                  0.0                NaN          14.0            52.0   

   Bene_Avg_Risk_Scre  
0            1.462630  
1            1.691054  
2            0.727543  
3            0.946591  
4            0.801164  

[5 rows x 85 columns]

**Shape of the dataset**

In [6]:
partD_pd.shape

(1287454, 85)

**List all the columns in partD dataset.**

In [7]:
partD_pd.columns

Index(['PRSCRBR_NPI', 'Prscrbr_Last_Org_Name', 'Prscrbr_First_Name',
       'Prscrbr_MI', 'Prscrbr_Crdntls', 'Prscrbr_Gndr', 'Prscrbr_Ent_Cd',
       'Prscrbr_St1', 'Prscrbr_St2', 'Prscrbr_City', 'Prscrbr_State_Abrvtn',
       'Prscrbr_State_FIPS', 'Prscrbr_zip5', 'Prscrbr_RUCA',
       'Prscrbr_RUCA_Desc', 'Prscrbr_Cntry', 'Prscrbr_Type',
       'Prscrbr_Type_src', 'Tot_Clms', 'Tot_30day_Fills', 'Tot_Drug_Cst',
       'Tot_Day_Suply', 'Tot_Benes', 'GE65_Sprsn_Flag', 'GE65_Tot_Clms',
       'GE65_Tot_30day_Fills', 'GE65_Tot_Drug_Cst', 'GE65_Tot_Day_Suply',
       'GE65_Bene_Sprsn_Flag', 'GE65_Tot_Benes', 'Brnd_Sprsn_Flag',
       'Brnd_Tot_Clms', 'Brnd_Tot_Drug_Cst', 'Gnrc_Sprsn_Flag',
       'Gnrc_Tot_Clms', 'Gnrc_Tot_Drug_Cst', 'Othr_Sprsn_Flag',
       'Othr_Tot_Clms', 'Othr_Tot_Drug_Cst', 'MAPD_Sprsn_Flag',
       'MAPD_Tot_Clms', 'MAPD_Tot_Drug_Cst', 'PDP_Sprsn_Flag', 'PDP_Tot_Clms',
       'PDP_Tot_Drug_Cst', 'LIS_Sprsn_Flag', 'LIS_Tot_Clms', 'LIS_Drug_Cst',
       'NonLIS_Sprsn_

**Take all the columns that woud be used in Machine Learning problem.**

In [8]:
part_D_full_data = partD_pd.loc[:,['PRSCRBR_NPI',
                                    'Prscrbr_Type',
                                    'Tot_Drug_Cst',\
                                    'Tot_Clms',\
                                    'Tot_Day_Suply', 'Tot_Benes']]
part_D_full_data = part_D_full_data.rename(columns = {"PRSCRBR_NPI": "npi", 
                                            'Prscrbr_Type': "specialty_description",
                                            "Tot_Drug_Cst": "total_drug_cost",
                                            'Tot_Clms': "total_claim",
                                            "Tot_Day_Suply": "total_day_supply",
                                            "Tot_Benes": "total_beneficiaries"})
part_D_full_data.head(5)

npi    specialty_description  total_drug_cost  total_claim  \
0  1003000126        Internal Medicine         81553.44         1123   
1  1003000142           Anesthesiology         37841.04         1493   
2  1003000167                  Dentist           221.66           47   
3  1003000175                  Dentist           125.82           20   
4  1003000423  Obstetrics & Gynecology         20757.65          206   

   total_day_supply  total_beneficiaries  
0             74080                358.0  
1             43944                299.0  
2               455                 35.0  
3               161                 11.0  
4             10231                 66.0

**The next dataset: Data about the total payments of physicians in each NPIs in the United States**

In [9]:
PHYSICIAN_PAYMENTS_DIR = os.path.join(".", "datasets", "PGYR21_P063023", "OP_DTL_GNRL_PGYR2021_P06302023.csv")
# Number of rows to retrieve randomly
num_rows_to_read = 1000000  # Adjust this number as needed
physician_payment_data = pd.read_csv(PHYSICIAN_PAYMENTS_DIR, 
                                     nrows = num_rows_to_read)
#Read the first rows of the dataframe
physician_payment_data.head(5)

/var/folders/kg/981chym54f147n0pfd03h4dr0000gn/T/ipykernel_7930/660411992.py:4: DtypeWarning: Columns (4,8,10,15,17,18,20,26,27,32,33,34,35,39,46,47,48,51,52,53,54,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88) have mixed types. Specify dtype option on import or set low_memory=False.
  physician_payment_data = pd.read_csv(PHYSICIAN_PAYMENTS_DIR,


Change_Type       Covered_Recipient_Type  Teaching_Hospital_CCN  \
0   UNCHANGED  Covered Recipient Physician                    NaN   
1   UNCHANGED  Covered Recipient Physician                    NaN   
2   UNCHANGED  Covered Recipient Physician                    NaN   
3   UNCHANGED  Covered Recipient Physician                    NaN   
4   UNCHANGED  Covered Recipient Physician                    NaN   

   Teaching_Hospital_ID Teaching_Hospital_Name  Covered_Recipient_Profile_ID  \
0                   NaN                    NaN                       92058.0   
1                   NaN                    NaN                     1231338.0   
2                   NaN                    NaN                      992939.0   
3                   NaN                    NaN                      774776.0   
4                   NaN                    NaN                      106575.0   

   Covered_Recipient_NPI Covered_Recipient_First_Name  \
0           1.043218e+09                         Ahad   
1           1.487818e+09                       Arsham   
2           1.104822e+09                       Donato   
3           1.457355e+09                         Earl   
4           1.346417e+09                      Michael   

  Covered_Recipient_Middle_Name Covered_Recipient_Last_Name  ...  \
0                           NaN                   Mahootchi  ...   
1                           NaN                    Sheybani  ...   
2                             J                    Borrillo  ...   
3                           NaN                      Craven  ...   
4                             J                       Bauer  ...   

  Associated_Drug_or_Biological_NDC_4  \
0                                 NaN   
1                                 NaN   
2                                 NaN   
3                                 NaN   
4                                 NaN   

  Associated_Device_or_Medical_Supply_PDI_4 Covered_or_Noncovered_Indicator_5  \
0                                       NaN                               NaN   
1                                       NaN                               NaN   
2                                       NaN                               NaN   
3                                       NaN                               NaN   
4                                       NaN                               NaN   

  Indicate_Drug_or_Biological_or_Device_or_Medical_Supply_5  \
0                                                NaN          
1                                                NaN          
2                                                NaN          
3                                                NaN          
4                                                NaN          

  Product_Category_or_Therapeutic_Area_5  \
0                                    NaN   
1                                    NaN   
2                                    NaN   
3                                    NaN   
4                                    NaN   

  Name_of_Drug_or_Biological_or_Device_or_Medical_Supply_5  \
0                                                NaN         
1                                                NaN         
2                                                NaN         
3                                                NaN         
4                                                NaN         

  Associated_Drug_or_Biological_NDC_5  \
0                                 NaN   
1                                 NaN   
2                                 NaN   
3                                 NaN   
4                                 NaN   

  Associated_Device_or_Medical_Supply_PDI_5 Program_Year  \
0                                       NaN         2021   
1                                       NaN         2021   
2                                       NaN         2021   
3                                       NaN         2021   
4                                       NaN         2021

**Taking only relevant columns here: "npi" of the physician and the total payment for them.**

In [10]:
physician_data = physician_payment_data.loc[:, ["Covered_Recipient_NPI", 'Total_Amount_of_Payment_USDollars']]
physician_data = physician_data.rename(columns = {"Covered_Recipient_NPI": "npi", 
                                                  "Total_Amount_of_Payment_USDollars": "total_payment"})
physician_data.head(5)

npi  total_payment
0  1.043218e+09         2500.0
1  1.487818e+09         3000.0
2  1.104822e+09         8580.0
3  1.457355e+09          500.0
4  1.346417e+09        37000.0

**Dropping NaN values in physician payments data**

In [11]:
physician_data = physician_data.dropna(subset = ["npi"])

**Groupby to calculate the aggregate payment sum for each individual NPI**

In [12]:
physician_data["npi"] = physician_data["npi"].astype(int)
total_payment_sum = physician_data.groupby(["npi"])["total_payment"].sum()
physician_data["total_payment_sum"] = physician_data["npi"].map(total_payment_sum.to_dict())
physician_data = physician_data.loc[:, ["npi", "total_payment_sum"]]
physician_data = physician_data.drop_duplicates()
physician_data

npi  total_payment_sum
0       1043218118            2983.89
1       1487817995           36790.20
2       1104822170            8580.00
3       1457354821             500.00
4       1346417086           38925.00
...            ...                ...
999987  1639141732              22.34
999989  1689052748              11.36
999991  1841230646              26.51
999995  1255339909              12.71
999997  1235355330              84.49

[240237 rows x 2 columns]

**Merging the data in two dataset together, according to the NPI column.**

In [14]:
full_dataset = pd.merge(part_D_full_data, physician_data, on = "npi", how = "left")
full_dataset

npi    specialty_description  total_drug_cost  total_claim  \
0        1003000126        Internal Medicine         81553.44         1123   
1        1003000142           Anesthesiology         37841.04         1493   
2        1003000167                  Dentist           221.66           47   
3        1003000175                  Dentist           125.82           20   
4        1003000423  Obstetrics & Gynecology         20757.65          206   
...             ...                      ...              ...          ...   
1287449  1992999650                  Dentist           498.08           74   
1287450  1992999692               Pharmacist          2167.29           15   
1287451  1992999817      Orthopaedic Surgery           248.81           25   
1287452  1992999825           Otolaryngology          7712.09          206   
1287453  1992999874        Internal Medicine         16239.45          304   

         total_day_supply  total_beneficiaries  total_payment_sum  
0                   74080                358.0                NaN  
1                   43944                299.0                NaN  
2                     455                 35.0                NaN  
3                     161                 11.0                NaN  
4                   10231                 66.0              35.88  
...                   ...                  ...                ...  
1287449              1173                 30.0                NaN  
1287450               282                 13.0                NaN  
1287451               219                 17.0                NaN  
1287452              4636                102.0                NaN  
1287453              6964                103.0                NaN  

[1287454 rows x 7 columns]

**Next dataset: The list of excluded NPI in the United States, as of 2023. That is the list of NPIs that have been
caught with medical frauds and labeled as such by the United States medical authority.**

In [15]:
EXCLUSION_CSV_DIR = os.path.join(".", "datasets", "LEIE.csv")
exclusion_data = pd.read_csv(EXCLUSION_CSV_DIR)
#Read the first rows of the dataframe
exclusion_data.head(5)

/var/folders/kg/981chym54f147n0pfd03h4dr0000gn/T/ipykernel_7930/4010586603.py:2: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.
  exclusion_data = pd.read_csv(EXCLUSION_CSV_DIR)


LASTNAME FIRSTNAME MIDNAME                      BUSNAME         GENERAL  \
0      NaN       NaN     NaN    #1 MARKETING SERVICE, INC  OTHER BUSINESS   
1      NaN       NaN     NaN             1 BEST CARE, INC  OTHER BUSINESS   
2      NaN       NaN     NaN  101 FIRST CARE PHARMACY INC  OTHER BUSINESS   
3      NaN       NaN     NaN     14 LAWRENCE AVE PHARMACY        PHARMACY   
4      NaN       NaN     NaN     143 MEDICAL EQUIPMENT CO     DME COMPANY   

            SPECIALTY UPIN         NPI  DOB                        ADDRESS  \
0          SOBER HOME  NaN           0  NaN      239 BRIGHTON BEACH AVENUE   
1  HOME HEALTH AGENCY  NaN           0  NaN  2161 UNIVERSITY AVENUE W, STE   
2            PHARMACY  NaN  1972902351  NaN  C/O 609 W 191ST STREET, APT D   
3                 NaN  NaN           0  NaN             14 LAWRENCE AVENUE   
4        DME - OXYGEN  NaN           0  NaN               701 NW 36 AVENUE   

         CITY STATE    ZIP EXCLTYPE  EXCLDATE  REINDATE  WAIVERDATE WVRSTATE  
0    BROOKLYN    NY  11235   1128a1  20200319         0           0      NaN  
1  SAINT PAUL    MN  55114   1128b5  20230518         0           0      NaN  
2    NEW YORK    NY  10040   1128b8  20220320         0           0      NaN  
3   SMITHTOWN    NY  11787   1128a1  19880830         0           0      NaN  
4       MIAMI    FL  33125   1128b7  19970620         0           0      NaN

**Retriving all the fraudulent NPIs, label them as 1 in the target feature "is_fraud".**

In [16]:
#Get all the rows whose "NPI" column is not equal to 0, that would be the list of excluded NPIs.
npi_fraud_data = exclusion_data.loc[:, ["NPI", "EXCLTYPE"]].query('NPI !=0')
#Rename the column
npi_fraud_data = npi_fraud_data.rename(columns = {"NPI": "npi", "EXCLTYPE": "is_fraud"})
#Set the "is_fraud" column to 1.
npi_fraud_data['is_fraud'] = 1
#View the first rows of the resulting data
npi_fraud_data.head(5)

npi  is_fraud
2   1972902351         1
6   1922348218         1
26  1942476080         1
30  1275600959         1
33  1891731758         1

In [17]:
len(npi_fraud_data)

7060

**Continue merging**

In [18]:
full_dataset = pd.merge(full_dataset, npi_fraud_data, on = "npi", how = "left")
full_dataset

npi    specialty_description  total_drug_cost  total_claim  \
0        1003000126        Internal Medicine         81553.44         1123   
1        1003000142           Anesthesiology         37841.04         1493   
2        1003000167                  Dentist           221.66           47   
3        1003000175                  Dentist           125.82           20   
4        1003000423  Obstetrics & Gynecology         20757.65          206   
...             ...                      ...              ...          ...   
1287450  1992999650                  Dentist           498.08           74   
1287451  1992999692               Pharmacist          2167.29           15   
1287452  1992999817      Orthopaedic Surgery           248.81           25   
1287453  1992999825           Otolaryngology          7712.09          206   
1287454  1992999874        Internal Medicine         16239.45          304   

         total_day_supply  total_beneficiaries  total_payment_sum  is_fraud  
0                   74080                358.0                NaN       NaN  
1                   43944                299.0                NaN       NaN  
2                     455                 35.0                NaN       NaN  
3                     161                 11.0                NaN       NaN  
4                   10231                 66.0              35.88       NaN  
...                   ...                  ...                ...       ...  
1287450              1173                 30.0                NaN       NaN  
1287451               282                 13.0                NaN       NaN  
1287452               219                 17.0                NaN       NaN  
1287453              4636                102.0                NaN       NaN  
1287454              6964                103.0                NaN       NaN  

[1287455 rows x 8 columns]

**Filling NaN values with 0, in here the rows whose column "is_fraud" is NaN would be filled with the value 0.**

In [19]:
full_dataset.fillna({"is_fraud": 0}, inplace = True)

In [20]:
full_dataset.head()

npi    specialty_description  total_drug_cost  total_claim  \
0  1003000126        Internal Medicine         81553.44         1123   
1  1003000142           Anesthesiology         37841.04         1493   
2  1003000167                  Dentist           221.66           47   
3  1003000175                  Dentist           125.82           20   
4  1003000423  Obstetrics & Gynecology         20757.65          206   

   total_day_supply  total_beneficiaries  total_payment_sum  is_fraud  
0             74080                358.0                NaN       0.0  
1             43944                299.0                NaN       0.0  
2               455                 35.0                NaN       0.0  
3               161                 11.0                NaN       0.0  
4             10231                 66.0              35.88       0.0

**The column "total_payment_sum" NaN values will be filled by the median value of column**

In [21]:
full_dataset['total_payment_sum'] = full_dataset['total_payment_sum'].fillna(
    full_dataset["total_payment_sum"].median())

In [22]:
full_dataset.head()

npi    specialty_description  total_drug_cost  total_claim  \
0  1003000126        Internal Medicine         81553.44         1123   
1  1003000142           Anesthesiology         37841.04         1493   
2  1003000167                  Dentist           221.66           47   
3  1003000175                  Dentist           125.82           20   
4  1003000423  Obstetrics & Gynecology         20757.65          206   

   total_day_supply  total_beneficiaries  total_payment_sum  is_fraud  
0             74080                358.0              55.57       0.0  
1             43944                299.0              55.57       0.0  
2               455                 35.0              55.57       0.0  
3               161                 11.0              55.57       0.0  
4             10231                 66.0              35.88       0.0

**Only 304 observations in the whole dataset are positive(there are fraud). The rest were negative(no fraud).**

In [23]:
full_dataset[full_dataset['is_fraud'] == 1].count()

npi                      304
specialty_description    304
total_drug_cost          304
total_claim              304
total_day_supply         304
total_beneficiaries      257
total_payment_sum        304
is_fraud                 304
dtype: int64

In [24]:
partD_pd.head(5)

PRSCRBR_NPI Prscrbr_Last_Org_Name Prscrbr_First_Name Prscrbr_MI  \
0   1003000126             Enkeshafi            Ardalan        NaN   
1   1003000142                Khalil             Rashid        NaN   
2   1003000167               Escobar              Julio          E   
3   1003000175         Reyes-Vasquez            Belinda        NaN   
4   1003000423               Velotta           Jennifer          A   

  Prscrbr_Crdntls Prscrbr_Gndr Prscrbr_Ent_Cd                 Prscrbr_St1  \
0            M.D.            M              I   6410 Rockledge Dr Ste 304   
1            M.D.            M              I  4126 N Holland Sylvania Rd   
2             DDS            M              I              5 Pine Cone Rd   
3          D.D.S.            F              I     322 N Azusa Ave Ste 202   
4            M.D.            F              I            11100 Euclid Ave   

  Prscrbr_St2 Prscrbr_City  ... Bene_Male_Cnt  Bene_Race_Wht_Cnt  \
0         NaN     Bethesda  ...         153.0              231.0   
1   Suite 220       Toledo  ...         105.0              144.0   
2         NaN       Dayton  ...          14.0               32.0   
3         NaN    La Puente  ...           NaN                NaN   
4         NaN    Cleveland  ...           NaN               57.0   

   Bene_Race_Black_Cnt  Bene_Race_Api_Cnt Bene_Race_Hspnc_Cnt  \
0                 32.0               47.0                20.0   
1                143.0                NaN                 NaN   
2                  0.0                0.0                 NaN   
3                  0.0                NaN                 NaN   
4                  NaN                0.0                 NaN   

  Bene_Race_Natind_Cnt Bene_Race_Othr_Cnt Bene_Dual_Cnt  Bene_Ndual_Cnt  \
0                  0.0               28.0         157.0           201.0   
1                  0.0                NaN         166.0           133.0   
2                  0.0                NaN           0.0            35.0   
3                  0.0                0.0           NaN             NaN   
4                  0.0                NaN          14.0            52.0   

   Bene_Avg_Risk_Scre  
0            1.462630  
1            1.691054  
2            0.727543  
3            0.946591  
4            0.801164  

[5 rows x 85 columns]

**Here I will pick more features from the first dataset to used for training ML models later.**

**The features that I add contains informations about the Prescriber(NPI) that provide medications and drugs(credentials, states,...);insurance claims, drug costs, total supply and total beneficiaries breakdown by brand-name drugs, generic drugs, opioid drugs, Medicare Advantage Prescription Drug (MAPD) and stand-alone Prescription Drug Plans (PDP), antibiotics; plus also demographic informations about the beneficiaries(ages, gender, average risk score).**

In [25]:
#All the numerical features from PartDPrescriberDataset that will be taken as feeatures used for Machine Learning
partD_pd_cat_features = ["Prscrbr_Crdntls", "Prscrbr_Gndr", "Prscrbr_State_Abrvtn", "Prscrbr_RUCA_Desc"]
#All the categorical features from PartDPrescriberDataset that will be taken as feeatures used for Machine Learning
partD_pd_num_features = ["GE65_Tot_Clms", "GE65_Tot_Drug_Cst", "GE65_Tot_Day_Suply", 'GE65_Tot_Benes',
                        "Brnd_Tot_Clms", "Brnd_Tot_Drug_Cst", "Gnrc_Tot_Clms", "Gnrc_Tot_Drug_Cst", 
                        "MAPD_Tot_Clms", "MAPD_Tot_Drug_Cst", "PDP_Tot_Clms", "PDP_Tot_Drug_Cst",
                        "LIS_Tot_Clms", "LIS_Drug_Cst", "NonLIS_Tot_Clms", "NonLIS_Drug_Cst", 
                        "Opioid_Tot_Clms", "Opioid_Tot_Drug_Cst", "Opioid_Prscrbr_Rate", 'Opioid_Tot_Benes',
                        'Opioid_LA_Tot_Clms', 'Opioid_LA_Tot_Drug_Cst', 'Opioid_LA_Tot_Suply', 'Opioid_LA_Tot_Benes', 
                        'Opioid_LA_Prscrbr_Rate',
                        "Antbtc_Tot_Clms", "Antbtc_Tot_Drug_Cst",
                        'Bene_Avg_Age', 'Bene_Age_LT_65_Cnt', 'Bene_Age_65_74_Cnt',
                        'Bene_Age_75_84_Cnt', 'Bene_Age_GT_84_Cnt', 'Bene_Feml_Cnt',
                        'Bene_Male_Cnt', 'Bene_Avg_Risk_Scre']

partD_pd_all_features = partD_pd_cat_features + partD_pd_num_features

#The subset of the original PartDPrescriberDataset that will be used for Machine Learning.
partD_data = partD_pd.loc[:, ["PRSCRBR_NPI"] + partD_pd_all_features]

**Continue merging**

In [26]:
full_dataset = pd.merge(full_dataset, partD_data, left_on = ["npi"], right_on = ["PRSCRBR_NPI"], 
                       how = "left")

In [27]:
len(full_dataset)

1287455

**Now we have the full dataset used for training**

In [28]:
full_dataset.drop(labels = ["PRSCRBR_NPI"], axis = 1, inplace = True)
full_dataset.head(5)

npi    specialty_description  total_drug_cost  total_claim  \
0  1003000126        Internal Medicine         81553.44         1123   
1  1003000142           Anesthesiology         37841.04         1493   
2  1003000167                  Dentist           221.66           47   
3  1003000175                  Dentist           125.82           20   
4  1003000423  Obstetrics & Gynecology         20757.65          206   

   total_day_supply  total_beneficiaries  total_payment_sum  is_fraud  \
0             74080                358.0              55.57       0.0   
1             43944                299.0              55.57       0.0   
2               455                 35.0              55.57       0.0   
3               161                 11.0              55.57       0.0   
4             10231                 66.0              35.88       0.0   

  Prscrbr_Crdntls Prscrbr_Gndr  ... Antbtc_Tot_Clms Antbtc_Tot_Drug_Cst  \
0            M.D.            M  ...            59.0              729.44   
1            M.D.            M  ...             NaN                 NaN   
2             DDS            M  ...            35.0              116.76   
3          D.D.S.            F  ...            15.0               93.09   
4            M.D.            F  ...             NaN                 NaN   

   Bene_Avg_Age  Bene_Age_LT_65_Cnt  Bene_Age_65_74_Cnt  Bene_Age_75_84_Cnt  \
0     76.435754                27.0               115.0               140.0   
1     65.381271               124.0               120.0                 NaN   
2     74.571429                 NaN                21.0                13.0   
3     70.818182                 NaN                 NaN                 NaN   
4     67.015152                 NaN                34.0                17.0   

   Bene_Age_GT_84_Cnt  Bene_Feml_Cnt  Bene_Male_Cnt  Bene_Avg_Risk_Scre  
0                76.0          205.0          153.0            1.462630  
1                 NaN          194.0          105.0            1.691054  
2                 NaN           21.0           14.0            0.727543  
3                 NaN            NaN            NaN            0.946591  
4                 NaN            NaN            NaN            0.801164  

[5 rows x 47 columns]

**Scaling numerical features to smaller number**

In [29]:
#all categorical features in the full dataset
all_cat_features = ["specialty_description"] + partD_pd_cat_features
#all numerical features in the full dataset
all_num_features = ["total_drug_cost", "total_claim", "total_day_supply", "total_payment_sum", 
                    "total_beneficiaries"] + partD_pd_num_features
#scaling of all numerical features
for feature in all_num_features:
    
    full_dataset[feature] = full_dataset[feature].map(lambda x: np.log10(x + 1.0))

full_dataset.head(5)

npi    specialty_description  total_drug_cost  total_claim  \
0  1003000126        Internal Medicine         4.911448     3.050766   
1  1003000142           Anesthesiology         4.577975     3.174351   
2  1003000167                  Dentist         2.347642     1.681241   
3  1003000175                  Dentist         2.103188     1.322219   
4  1003000423  Obstetrics & Gynecology         4.317199     2.315970   

   total_day_supply  total_beneficiaries  total_payment_sum  is_fraud  \
0          4.869707             2.555094           1.752586       0.0   
1          4.642909             2.477121           1.752586       0.0   
2          2.658965             1.556303           1.752586       0.0   
3          2.209515             1.079181           1.752586       0.0   
4          4.009961             1.826075           1.566791       0.0   

  Prscrbr_Crdntls Prscrbr_Gndr  ... Antbtc_Tot_Clms Antbtc_Tot_Drug_Cst  \
0            M.D.            M  ...        1.778151            2.863585   
1            M.D.            M  ...             NaN                 NaN   
2             DDS            M  ...        1.556303            2.070998   
3          D.D.S.            F  ...        1.204120            1.973543   
4            M.D.            F  ...             NaN                 NaN   

   Bene_Avg_Age  Bene_Age_LT_65_Cnt  Bene_Age_65_74_Cnt  Bene_Age_75_84_Cnt  \
0      1.888942            1.447158            2.064458            2.149219   
1      1.822046            2.096910            2.082785                 NaN   
2      1.878358                 NaN            1.342423            1.146128   
3      1.856234                 NaN                 NaN                 NaN   
4      1.832606                 NaN            1.544068            1.255273   

   Bene_Age_GT_84_Cnt  Bene_Feml_Cnt  Bene_Male_Cnt  Bene_Avg_Risk_Scre  
0            1.886491       2.313867       2.187521            0.391399  
1                 NaN       2.290035       2.025306            0.429922  
2                 NaN       1.342423       1.176091            0.237429  
3                 NaN            NaN            NaN            0.289275  
4                 NaN            NaN            NaN            0.255553  

[5 rows x 47 columns]

**Split the full dataset into features and targets.**

In [30]:
#Split the full dataset into features and targets.
X_full_dataset = full_dataset.drop(["is_fraud"], axis = 1)
Y_full_dataset = full_dataset["is_fraud"].astype(int).copy()

Y_full_dataset

0          0
1          0
2          0
3          0
4          0
          ..
1287450    0
1287451    0
1287452    0
1287453    0
1287454    0
Name: is_fraud, Length: 1287455, dtype: int64

In [31]:
Y_full_dataset.unique()

array([0, 1])

In [32]:
X_full_dataset.head(3)

npi specialty_description  total_drug_cost  total_claim  \
0  1003000126     Internal Medicine         4.911448     3.050766   
1  1003000142        Anesthesiology         4.577975     3.174351   
2  1003000167               Dentist         2.347642     1.681241   

   total_day_supply  total_beneficiaries  total_payment_sum Prscrbr_Crdntls  \
0          4.869707             2.555094           1.752586            M.D.   
1          4.642909             2.477121           1.752586            M.D.   
2          2.658965             1.556303           1.752586             DDS   

  Prscrbr_Gndr Prscrbr_State_Abrvtn  ... Antbtc_Tot_Clms  Antbtc_Tot_Drug_Cst  \
0            M                   MD  ...        1.778151             2.863585   
1            M                   OH  ...             NaN                  NaN   
2            M                   NV  ...        1.556303             2.070998   

   Bene_Avg_Age  Bene_Age_LT_65_Cnt  Bene_Age_65_74_Cnt  Bene_Age_75_84_Cnt  \
0      1.888942            1.447158            2.064458            2.149219   
1      1.822046            2.096910            2.082785                 NaN   
2      1.878358                 NaN            1.342423            1.146128   

   Bene_Age_GT_84_Cnt  Bene_Feml_Cnt  Bene_Male_Cnt  Bene_Avg_Risk_Scre  
0            1.886491       2.313867       2.187521            0.391399  
1                 NaN       2.290035       2.025306            0.429922  
2                 NaN       1.342423       1.176091            0.237429  

[3 rows x 46 columns]

**There are more than 1 million instances in the full dataset, while there are only 304 positive instances in the whole dataset. So the dataset is grossly imbalanced, so here I use RandomUnderSampler in the library imblearn to choose randomly 80000 negative samples from more than 80000 negative samples for the full dataset along with 200 positive instances.**

In [33]:
from imblearn.under_sampling import RandomUnderSampler
# Specify the fixed number of negative instances you want to keep
num_negative_samples = 80000
# Create a RandomUnderSampler
rus = RandomUnderSampler(sampling_strategy = {0: num_negative_samples, 1: 200}, random_state = 41)

# Fit and transform the training data using the RandomUnderSampler
X_resampled_full_dataset, y_resampled_full_dataset = rus.fit_resample(X_full_dataset, Y_full_dataset)

In [34]:
len(X_resampled_full_dataset)

80200

**Divide the dataset into train set and test set. I will divide the data in a stratified way, to ensure that the positive/negative ratio in train set and test set is identical.**

In [35]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X_resampled_full_dataset, 
                y_resampled_full_dataset, test_size = 0.25, random_state=42, stratify = y_resampled_full_dataset)

In [36]:
y_train[y_train == 1].sum(), y_test[y_test == 1].sum()

(150, 50)

**Number of unique values in each categorical features**

In [37]:
for features in all_cat_features:
    print(len(X_train[features].unique()))

125
1895
2
59
16


# Preprocessing data in Scikit-Learn Pipeline

**Importing all relevant classes in sklearn.**

In [38]:
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import OneHotEncoder

In [39]:
from sklearn.compose import make_column_selector

**Creatng pipeline for numerical features and categorical features of the dataset.**

In [40]:
num_pipeline = Pipeline([("imputer", SimpleImputer(strategy = "constant", fill_value = 0)),
                        ("std_scaler", StandardScaler())])
cat_pipeline = Pipeline([("one_hot_encoder",
                OneHotEncoder(handle_unknown = "infrequent_if_exist", min_frequency = 10, max_categories = 10))])

**The ColumnTransformer which apply numerical pipeline to numerical columns and categorical pipeline to categorical columns, and then concatenating the results of those 2 transformations along the feature axis.**

In [41]:
col_transform = ColumnTransformer(transformers = [
    ("num_pipeline", num_pipeline, all_num_features),
    ("cat_pipeline", cat_pipeline, all_cat_features)], remainder = "drop", sparse_threshold = 0.0)

In [42]:
full_pipeline = Pipeline([("col_transform", col_transform)])
X_train_prepared = full_pipeline.fit_transform(X_train)

In [43]:
X_train_prepared.shape

(60150, 82)

# Training and testing with Kafka

**Importing Machine Learning training algorithms**

In [133]:
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import ExtraTreesClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import GradientBoostingClassifier

**Creating MachineLearning class**

In [134]:
log_reg = LogisticRegression(random_state = 42, class_weight = {0: 1, 1: 500}, max_iter = 400, tol = 1e-2)
random_forest = RandomForestClassifier(n_estimators = 300, random_state = 42,
                                       class_weight = {0: 1, 1: 500})
grad_boost = GradientBoostingClassifier(n_estimators = 200, random_state = 42)

classifiers = [log_reg, random_forest, grad_boost]

**Training each classifier in the list**

In [135]:
for classifier in classifiers:
    
    classifier.fit(X_train_prepared, y_train)

/Users/nguyentuannguyen/opt/anaconda3/envs/py310/lib/python3.10/site-packages/sklearn/linear_model/_logistic.py:444: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


**Testing**

In [136]:
from kafka import KafkaProducer
import json
import time
import random

In [137]:
X_test_prepared = full_pipeline.transform(X_test)

**Creating a topic named "testing" to send testing data into. Also creating a KafkaProducer to send data into the topic**

In [93]:
from kafka import KafkaProducer
# Kafka broker address, in this case will be locathost:9092
kafka_broker = 'localhost:9092'
# Kafka topic to publish data, in this case will be "training"
testing_kafka_topic = 'testing'

# Create a Kafka producer
testing_producer = KafkaProducer(bootstrap_servers = [kafka_broker], 
                         value_serializer = lambda v: json.dumps(v).encode('utf-8'), 
                        api_version = (0, 10, 1))


**Send data into the testing topic**

In [52]:
#Sending training instances to the newly created Kafka topic
for i in range(len(X_test_prepared)):
    
    # Combine random data with actual Iris dataset features
    instance = {
        'features': X_test_prepared[i].tolist(),
        'label': int(y_test.iloc[i])
    }

    # Publish data to Kafka topic
    testing_producer.send(testing_kafka_topic, value = instance)

# Close the Kafka producer
testing_producer.close()

**Creating KafkaConsumer to read data from the Kafka testing topic**

In [138]:
from kafka import KafkaConsumer
# Create a Kafka consumer
testing_consumer = KafkaConsumer(testing_kafka_topic, bootstrap_servers = [kafka_broker], 
                value_deserializer = lambda x: json.loads(x.decode('utf-8')), 
                                 consumer_timeout_ms = 30000, 
                                 auto_offset_reset = "earliest")

In [139]:
from sklearn.metrics import roc_auc_score
from sklearn.metrics import f1_score
from sklearn.metrics import precision_score
from sklearn.metrics import recall_score

import tensorflow as tf

**Now reading from Kafka Testing topic and calculate streaming metrics for each classifiers. For each 100 newly added samples in the testing set the streaming metrics will be calculated.**

In [140]:
features = []
labels = []

#I will use 3 metrics: Precision, Recall and AreaUnderTheCurve to 
#evaluate the efficiency of 3 different classifiers
metrics_list = [(classifier, (tf.keras.metrics.Precision(thresholds = 0.02), 
                                  tf.keras.metrics.Recall(thresholds = 0.02), 
                                  tf.keras.metrics.AUC()) ) for classifier in classifiers]
# Consume and process data
for message in testing_consumer:
    data_point = message.value
    
    # Collect features and labels
    features.append(data_point['features'])
    labels.append(data_point['label'])
    
    if len(features) >= 100:
        for classifier, (precision, recall, auc) in metrics_list:
            sample = np.array(features)
            label = np.array(labels, dtype = np.int16)
            precision.update_state(label, 
                                classifier.predict_proba(sample)[:, 1])
            recall.update_state(label, 
                                classifier.predict_proba(sample)[:, 1])
            auc.update_state(label, 
                                classifier.predict_proba(sample)[:, 1])
        
        # Reset data lists
        features = []
        labels = []

# Close the Kafka consumer
testing_consumer.close()

**Printing all 3 metrics for both 3 classfiers. It can be concluded that LogisticRegression will be the most effective classifier for the problem.**

In [141]:
for classifier, (precision, recall, auc) in metrics_list:
    
    print(classifier.__class__.__name__, "\n")
    print("The precision score on the test set is:", precision.result().numpy())
    print("The recall score on the test set is:", recall.result().numpy())
    print("The roc auc score on the test set is:", auc.result().numpy(), "\n")

LogisticRegression 

The precision score on the test set is: 0.0027059864
The recall score on the test set is: 0.98
The roc auc score on the test set is: 0.8011905 

RandomForestClassifier 

The precision score on the test set is: 0.01724138
The recall score on the test set is: 0.12
The roc auc score on the test set is: 0.6562482 

GradientBoostingClassifier 

The precision score on the test set is: 0.0
The recall score on the test set is: 0.0
The roc auc score on the test set is: 0.52317137 

